libyul/optimiser/UnusedFunctionReturnParameterPruner.h

```diff
@@ -0,0 +1,59 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/optimiser/OptimiserStep.h>
+
+namespace solidity::yul
+{
+
+/**
+ * UnusedFunctionReturnParameterPruner: Optimiser step that removes unused return parameters at function callsite.
+ *
+ * If a parameter is unused, like `z` and correspondingly `x` in the following snippet:
+ *
+ * {
+ *    let z := f(sload(1), sload(2))
+ *    function f(a,b) -> x { x := div(a,b) }
+ * }
+ *
+ * We remove the parameter and create a new "linking" function `f2` as follows:
+ *
+ * function f(a,b) {
+ *   let x
+ *   x := div(a,b)
+ * }
+ *
+ * function f2(a,b) -> x { x := f(a,b) }
+ *
+ * and replace all references to `f` by `f2`.
+ * The inliner should be run afterwards to make sure that all references to `f2` are replaced by
+ * `f`.
+ *
+ * Prerequisites: Disambiguator, FunctionHoister, ForLoopInitRewriter, SSATransform,
```

Comment:
Are all these needed for correctness?

Reply:
Disambiguator, FunctionHoister are needed like other steps.

ExpressionSplitter is needed for correctness, otherwise, the following example will be incorrectly pruned:

```solidity
// will be unused
x := f()
sstore(f(), 1)
````

Because of ExpressionSplitter, we need ForLoopConditionIntoBody.

SSATransform is useful in cases such as the following to realize that `x` is unused:

```solidity
let x := f()
x := 1
````

SSATransform has the prerequisite ForLoopinitRewriter.

I'll add this as a comment.

Reply:
I don't think it is a good idea to rely on all these for correctness. It will also not be compatible with the phaser or with users being able to specify an order of optimizer steps.